### PR TITLE
SBRA-739 redefine the graph creation method

### PR DESCRIPTION
**Pull request by A669015**
This PR refactor the code to avoid the computation of the graph topology for each sample.
The graph topology is now computed on the datamodule side at initialization, and passed to the ligthning module through the argument linking feature of ligthning. Then at training time, the dataloader only send the features to the model.

### CODE REVIEW

File: reactive-flows/cnf-combustion/gnns/data.py [hunk 1]
```diff
@@ -70,14 +73,38 @@ def download(self) -> None:
...
+        data = self._get_data(idx)
+        pyg_data.x = tensor(data["feat"].reshape(-1, 1), dtype=tfloat)
+        pyg_data.y = tensor(data["sigma"].reshape(-1, 1), dtype=tfloat)
+        return pyg_data
+
+    def create_graph_topo(self, grid_shape: Tuple[int, int, int]) -> None:
+        """Create the graph topology and store it in memory.
+
+        Args:
+            grid_shape (Tuple[int, int, int]): the shapr of the grid for the
```
Comment (elasto) [inline reactive-flows/cnf-combustion/gnns/data.py:100]: typo: shapr --> shape

**Review by elasto — APPROVED** [review]

File: reactive-flows/cnf-combustion/gnns/data.py [hunk 2]
```diff
@@ -71,14 +73,25 @@ def download(self) -> None:
...
 
         Returns:
-            (pyg.data.Data): Graph at the given index.
+            (Dict[str, np.array]): the feat and sigma.
         """
-        data = torch.load(os.path.join(self.processed_dir, f"data-{idx}.pt"))
-        return data
+        raise NotImplementedError
+
+    def get(self, idx: int) -> Dict[str, tensor]:
```
Comment (mikael10j) [inline reactive-flows/cnf-combustion/gnns/data.py:84]: The type hint isn't consistent with the returned object

**Review by mikael10j — COMMENTED** [review]
Just a quick review

**Review by mikael10j — APPROVED** [review]

File: reactive-flows/cnf-combustion/gnns/tests/test_data/test_model.ckpt
Comment (cbovalo) [inline reactive-flows/cnf-combustion/gnns/tests/test_data/test_model.ckpt:1]: Should we keep this checkpoint file ?

**Comment by cbovalo** [comment]
Just a few comments:
* in the `pyg_data` object, I suggest setting `edge_index` to None to make it clearer that the adjacency matrix will be given later by `graph_topology`,
* I think the `egde_index` should be a buffer of the `nn.Module` to be put in the correct device automatically (right now it stays on cpu),
* I try to think to a better way to pass the grid shape (for plotting) instead of using the position. I mean do we need a tensor of shape [num_nodes, num_dimensions] to get only 3 numbers ?

File: reactive-flows/cnf-combustion/gnns/tests/test_data/test_model.ckpt
Comment (A669015) [reply reactive-flows/cnf-combustion/gnns/tests/test_data/test_model.ckpt:1]: Yes this model is still used in the tests

**Comment by A669015** [comment]
> * I think the `egde_index` should be a buffer of the `nn.Module` to be put in the correct device automatically (right now it stays on cpu),

Do you suggest to register the `edge_index` tensor in the `state_dict`, so in the model checkpoint ? In this case the graph topology creation will no more be usefull at inference time, but the model will be usable only for the trained data type.

> * I try to think to a better way to pass the grid shape (for plotting) instead of using the position. I mean do we need a tensor of shape [num_nodes, num_dimensions] to get only 3 numbers ?

What do you mean by "a tensor of shape [num_nodes, num_dimensions] to get only 3 numbers" ? The `grid_shape` is a tuple of 3 values.

**Comment by cbovalo** [comment]
> > * I think the `egde_index` should be a buffer of the `nn.Module` to be put in the correct device automatically (right now it stays on cpu),
> 
> Do you suggest to register the `edge_index` tensor in the `state_dict`, so in the model checkpoint ? We can make it non-persistent to keep the checkpoint as light as possible.
> 
Not in the `state_dict` (setting `persistent=False` in `register_buffer`)
> > * I try to think to a better way to pass the grid shape (for plotting) instead of using the position. I mean do we need a tensor of shape [num_nodes, num_dimensions] to get only 3 numbers ?
> 
> What do you mean by "a tensor of shape [num_nodes, num_dimensions] to get only 3 numbers" ? The `grid_shape` is a tuple of 3 values.
Exactly! What I mean is that you add to `graph_topology` an attribute `pos` which is a tensor with shape (num_nodes, 3) - Cartesian coordinates in 3D. You retrieve the min/max in each axis to get the domain size. Maybe this shape could be returned by the Datamodule and be liked to the LightningModule.
